만국박람회 [STEP 1] 재재, 보리사랑

ZZBAE opened the pull request:
안녕하세요 제이슨! @ehgud0670
만국박람회 프로젝트를 같이 하게 된 재재와 보리사랑입니다! 잘 부탁드립니다 🙇🏻‍♀️

## ✅ 공부한 부분
- [x] JSON 의 기초개념과 이번 만국박람회 프로젝트에서 왜 사용되는지 활동학습을 통해 공부해보았습니다.
- [x] Codable을 채택하여 JSON 데이터와 매칭할 모델 타입을 구현하였습니다.
- [x] 스네이크 케이스인 JSON 키 값을 스위프트의 네이밍에 맞게 변환해주었습니다.

## ⁉️ 궁금한 점
- 소스 파일들을 어떤 폴더로 정리하시는지 궁금합니다. MVC를 예로 들면 모델끼리, 뷰끼리, 컨트롤러끼리 묶는지,아니면 한 뷰 컨트롤러가 담당하는 화면별로 MVC를 하나로 묶는지, 어떤 방식으로 파일을 계층화해서 정리하시는지 궁금합니다.
- 포크해서 클론받은 작업물 중에 `items.json`을 구현해주었는데, `Contents.json`도 해주어야하는지 의문이였습니다. 해주지 않아도 된다면 이 `Contents.json`은 어떤 역할을 하는 건지 궁금합니다.

ZZBAE committed:
feat: ExpoItemElement에 Codable 채택 후 매칭 모델 타입 구현

yusw10 committed:
feat: Exposition 구조체의 정의

- title, visitors, location, duration, expositionDescription 프로퍼티 선언
- 각 프로퍼티 별 JSON타입으로 디코딩하기 위해 CodingKey설정

ZZBAE committed:
refactor: 파일 Description 수정

- 작성자 팀원 이름으로 수정

ehgud0670 commented on Expo1900/Expo1900/Models/ExpoItemElement.swift:
```diff
@@ -0,0 +1,22 @@
+
+// MARK: - ExpoItemElement
+struct ExpoItemElement: Codable {
+    let name, imageName, shortDesc, desc: String
```
```suggestion
    let name: String
    let imageName: String
    let shortDesc: String
    let desc: String
```
=> 프로퍼티를 쉼표로 연달아 선언하는 것 보단 하나씩 선언하는게 파악하기도 쉽고 차후에 변경하기에도 좋다고 생각합니다. 
어떻게 생각하실지 궁금하네요 :)

ZZBAE replied:
개인적으로 한 줄로 쓰면 보기 편하겠다 생각했었는데, 차후에 변경하기 좋고 오히려 하나씩 분리해서 선언하는게 파악하기 쉽다는 제이슨의 의견에 따라 변경해주었습니다. 감사합니다 !

ehgud0670 commented on Expo1900/Expo1900/Models/ExpoItemElement.swift:
```diff
@@ -0,0 +1,22 @@
+struct ExpoItemElement: Codable {
+    let name, imageName, shortDesc, desc: String
+
+    enum CodingKeys: String, CodingKey {
+        case name
+        case imageName = "image_name"
+        case shortDesc = "short_desc"
+        case desc
+    }
```
CodingKeys 잘 사용하셨네요!! 
다른 방법으로는 JSONDecoder의 keyDecodingStrategy 중 `convertFromSnakeCase` 를 사용하는 방법도 있습니다. 
지금 구현한 방법도 좋으니, 안 바꾸셔도 되고 그냥 참고차 말씀드립니다~ :)

ZZBAE replied:
다른 방법들 알려주신 것 토대로 개인공부해보겠습니다. 🫡

ehgud0670 commented on Expo1900/Expo1900/Models/ExpoItemElement.swift:
```diff
@@ -0,0 +1,22 @@
+    }
+}
+
+typealias ExpoItem = [ExpoItemElement]
```
ExpoItem은 typealias로 선언만 하고 안 쓰는 것 같은데 선언을 지금 한 이유가 있으실까요~? 
PR를 올릴때 보통 선언을 했다면 쓰는 곳이 있을 거라 생각되어서요. 그 다음 PR에서 해당 타입이 쓰인다면 지금은 지우고 나중에 선언하시는 게 나을 듯 합니다 :)
아니면 이번 PR에서 테스트 코드라도 만들어서 실제로 json 이 모델로 잘 변환되는지 테스트 해보시면 납득이 될 듯 합니다! :)

yusw10 replied:
STEP2를 읽으면서 진행하다 보니 컨텐츠요소들을 배열로 묶어서 사용할 것 같아서 미리 작성했던것 같습니다...! 이 부분도 현재는 제거하였고 후에 STEP2 진행하면서 사용하도록 하겠습니다.
한가지 궁금한게 제가 지난번 계산기 프로젝트때 인스턴스 끼리는 같은지 테스트코드에서 비교가 안되어서 현재 작성한 테스트 코드에서는

1. json파일의 데이터를 `ExpoItemElement` 타입의 배열로 디코딩
2. `ExpoItemElement`의 배열 중 인스턴스 한개를 추출
3. 실제 직접 만든 인스턴스와 비교

이런식으로 비교를 해서 테스트를 진행했는데(내부 프로퍼티를 직접 비교하는 것 까지는 잘 된 것 같습니다) 
혹시 테스트 코드에서 인스턴스가 동일한지 비교할 때 `Equtable`을 채택해서 해야 하는건지, 아니면 다른 방법이 있는건지 궁금합니다.
`XCTAssertEqual(DecodedData, [ExpoItemElement]) // DecodedData.self 는[ExpoItemElement] 타입 `
이런식으로 인스턴스끼리의 직접비교가 불가능 한 것 같아서 여쭤봅니다..!

[Typealias 제거](https://github.com/yagom-academy/ios-exposition-universelle/pull/166/commits/6e0e297c7dc0af2cbb8827d523b872b048a3c483)
[테스트코드 작성](https://github.com/yagom-academy/ios-exposition-universelle/pull/166/commits/23e7d57521f4b50ccf0b95708c85828795390191)

ehgud0670 replied:
> json파일의 데이터를 ExpoItemElement 타입의 배열로 디코딩
ExpoItemElement의 배열 중 인스턴스 한개를 추출
실제 직접 만든 인스턴스와 비교

이 방법은 실제로 만든 인스턴스와 디코딩해서 가지고 온 인스턴스와 결국에는 같은 인스턴스가 아니라서 맞는 테스트는 아니라고 생각되네요! 단순히 디코딩 했을 때 해당 모델이 nil인지 아닌지(`try?` 로 디코딩한다면) 검사하면 될 듯 합니다. 그럼 결국 잘 파싱이 된거고 디코딩 테스트 통과라서요.
고민해보세요!!

ehgud0670 commented on Expo1900/Expo1900/Models/Exposition.swift:
```diff
@@ -0,0 +1,20 @@
+struct Exposition: Codable {
+    let title: String
+    let visitors: Int
+    let location, duration, expositionDescription: String
+
+    enum CodingKeys: String, CodingKey {
+        case title, visitors, location, duration
+        case expositionDescription = "description"
```
여기도 하나씩 따로 선언하는게 나을 듯 한데 어떻게 생각하실지 궁금하네용
연달아 프로퍼티를 선언한 이유가 있으실까요~? 🧐

yusw10 replied:
`CodingKey` 부분도 매핑을 해주는 부분과 해주지 않는 부분을 분리해서 표현하면 읽기 편할것같다고 생각했었습니다..!
다만 작성할 당시에는 `title`을 비롯한 매핑에 딱히 변경이 없는 프로퍼티들은 묶어서 놓는게 읽기 편할것 같다고 생각했는데 말씀주신대로 따로 놓고 보니 읽기 더 편했던것 같습니다.

[dcffaa2](https://github.com/yagom-academy/ios-exposition-universelle/pull/166/commits/dcffaa21e06253b365a3613d6ecf5951e0fcbe21)

ehgud0670 replied:
네네 만약 서버쪽에서 response 의 key값이 바뀔 경우도 고려하여 따로 위치하는게 좋기는 한 것 같아요. 수고하셨습니다 :)

ehgud0670 reviewed (changes requested):
테스트 코드로 JSON 데이터가 모델로 잘 변환되는지 이번 PR에서 꼭 반영해주세요. 
그래야 제가 리뷰어로서 어떤 데이터를 파싱하시는지 좀 더 파악이 될 것 같고 이해가 될 것 같아요.
그리고 여러분도 여러분이 작성하신 모델이 잘 파싱되는지도 확신할 수 있어서 안심이 되실 겁니다. 
그럼 피드백 참고 부탁드릴게요~ 😄

ehgud0670 commented on Expo1900/Expo1900/Models/ExpositionUniverselle.swift:
```diff
@@ -0,0 +1,8 @@
+//
```
이 파일은 아직 코드가 없다면 PR 올리기전에 삭제하면 될 것 같아요 :)

ZZBAE replied:
제가 파일을 새로 만들다가 remove 해주었는데 육안으로 봤을 때 xcode상에서는 지워져서 파일도 없어진 줄 알았습니다. 실수로 인한 잉여파일이여서 바로 제거해주었습니다!

ehgud0670 commented on Expo1900/Expo1900/Models/Exposition.swift:
```diff
@@ -0,0 +1,20 @@
+
+// MARK: - Exposition
+struct Exposition: Codable {
+    let title: String
+    let visitors: Int
+    let location, duration, expositionDescription: String
```
네, 해당 프로퍼티들은 화면에 보이기만 하는 데이터들이라 var(mutable) 보단 let(immutable) 이 더 적절하다고 생각합니다.
잘 하셨어요!! 👏👏👏

ZZBAE replied:
감사합니다 😆

ZZBAE committed:
refactor: ExpoItemElement 프로퍼티 분리

- 리뷰 반영하여 프로퍼티를 묶어서가 아닌 각각 따로 선언을 해주었음

ZZBAE committed:
refactor: ExpositionUniverselle 파일 제거

- 사용하지 않는 파일 제거해주었습니다.

yusw10 committed:
refactor: Exposition의 ₩프로퍼티 분리

- 리뷰 반영하여 구조체의 내부 프로퍼티를 각각 따로 선언을 해주었음
- 구조체의 CodingKey enum에서 case표현을 각각 따로 선언
- 타입별로 묶어서 표현하기 위해 visitors 프로퍼티 위치 변경

yusw10 committed:
refactor: Exposition의 프로퍼티 분리

- 리뷰 반영하여 구조체의 내부 프로퍼티를 각각 따로 선언을 해주었음
- 구조체의 CodingKey enum에서 case표현을 각각 따로 선언
- 타입별로 묶어서 표현하기 위해 visitors 프로퍼티 위치 변경

yusw10 committed:
Merge branch 'step1' of https://github.com/ZZBAE/ios-exposition-universelle into step1

ZZBAE committed:
test: JSON의 decoding test 코드 작성

- JSON이 모델로 변환이 잘 되는지 확인할 수 있는 test 코드 작성

yusw10 committed:
test: ExpoItemElement의 테스트코드 수정

- 불필요한 do-catch문 제거
- 인스턴스 비교를 위한 테스트코드 추가
- JSONDecoder를 setup 및 tearDown 메서드에서 인스턴스화

ehgud0670 commented:
# 궁금한 점

> 소스 파일들을 어떤 폴더로 정리하시는지 궁금합니다. MVC를 예로 들면 모델끼리, 뷰끼리, 컨트롤러끼리 묶는지,아니면 한 뷰 컨트롤러가 담당하는 화면별로 MVC를 하나로 묶는지, 어떤 방식으로 파일을 계층화해서 정리하시는지 궁금합니다.

=> 말씀주신것처럼 모델, 뷰, 컨트롤러끼리 묶는 경우도 있고 화면별로 모델, 뷰, 컨트롤러 폴더를 따로 묶는 경우도 있습니다.
화면이 너무 많아지다 보면 후자가 좀 더 파일 파악하기는 수월하더라고요. 그럼 고민해보시고 그룹 잘 나눠보시기 바랍니다 :)

>   포크해서 클론받은 작업물 중에 items.json을 구현해주었는데, Contents.json도 해주어야하는지 의문이였습니다. 해주지 않아도 된다면 이 Contents.json은 어떤 역할을 하는 건지 궁금합니다.

Contents.json 도 결국 화면에서 쓰인다면 같이 모델링 작업 해주시면 좋습니다. 저의 경우에는 서버 api 로 내려오는 response들을 하나의 브랜치로 모두 처리했었는데 이러면 한꺼번에 모델링 처리하는 거라 나중에 다시 모델링 작업을 해줄 필요가 없어 좋았습니다. 따라서 같이 작업해주시면 좋을 듯 하네용 😄

ehgud0670 commented on Expo1900/Expo1900/Models/Exposition.swift:
```diff
@@ -0,0 +1,25 @@
+import Foundation
+
+// MARK: - Exposition
+struct Exposition: Codable {
```
Codable 은 Encodable, Decodable을 모두 포함하는 프로토콜입니다. 
따라서 디코딩만 하는 모델이라면 Codable이 아닌 Decodable 로 범위 축소해서 채택시키는 것이 좋긴 해요.
코드를 볼 때 해당 모델은 디코딩만 하는 모델이라고 파악이 되기 때문입니다(내 동료도 디코딩만 하는 객체라고 생각할 수 있어서 좋아요).
그럼 참고 부탁드릴게요~!
